File: configuration/src/test/java/com/synopsys/integration/configuration/config/MultiplePropertySourceTests.java
package com.synopsys.integration.configuration.config;

import static com.synopsys.integration.configuration.util.ConfigTestUtils.configOf;
import static com.synopsys.integration.configuration.util.ConfigTestUtils.propertySourceOf;

import java.util.Optional;

import org.apache.commons.lang3.tuple.Pair;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.synopsys.integration.common.util.Bds;
import com.synopsys.integration.configuration.property.base.NullableAlikeProperty;
import com.synopsys.integration.configuration.property.types.string.NullableStringProperty;
import com.synopsys.integration.configuration.source.PropertySource;

class MultiplePropertySourceTests {
    @Test
    public void primaryValueUsedOverSecondary() throws InvalidPropertyException {
        NullableAlikeProperty<String> sharedProperty = new NullableStringProperty("shared.key");
        PropertySource secondarySource = propertySourceOf("secondaryName", Pair.of(sharedProperty.getKey(), "secondaryValue"));
        PropertySource primarySource = propertySourceOf("primaryName", Pair.of(sharedProperty.getKey(), "primaryValue"));

        PropertyConfiguration config = configOf(primarySource, secondarySource);
        Assertions.assertEquals(Optional.of("primaryValue"), config.getValue(sharedProperty));
        Assertions.assertEquals(Optional.of("primaryName"), config.getPropertySource(sharedProperty));
    }

    @Test
    public void valueFromSecondaryWhenNotInPrimary() throws InvalidPropertyException {
        NullableAlikeProperty<String> property = new NullableStringProperty("any.key");
        PropertySource secondarySource = propertySourceOf("secondaryName", Pair.of(property.getKey(), "secondaryValue"));
        PropertySource primarySource = propertySourceOf("primaryName");

        PropertyConfiguration config = configOf(primarySource, secondarySource);
        Assertions.assertEquals(Optional.of("secondaryValue"), config.getValue(property));
        Assertions.assertEquals(Optional.of("secondaryName"), config.getPropertySource(property));
    }

    @Test
    public void containsKeysFromBothSources() {
        NullableAlikeProperty<String> primaryProperty = new NullableStringProperty("primary.key");
        PropertySource primarySource = propertySourceOf("primaryName", Pair.of(primaryProperty.getKey(), "primaryValue"));

        NullableAlikeProperty<String> secondaryProperty = new NullableStringProperty("secondary.key");
        PropertySource secondarySource = propertySourceOf("secondaryName", Pair.of(secondaryProperty.getKey(), "secondaryValue"));

        PropertyConfiguration config = configOf(primarySource, secondarySource);
        Assertions.assertEquals(Bds.setOf(primaryProperty.getKey(), secondaryProperty.getKey()), config.getKeys());
    }

    @Test
    public void rawValueMapContainsValuesFromBothSources() {
        NullableAlikeProperty<String> primaryProperty = new NullableStringProperty("primary.key");
        PropertySource primarySource = propertySourceOf("primaryName", Pair.of(primaryProperty.getKey(), "primaryValue"));

        NullableAlikeProperty<String> secondaryProperty = new NullableStringProperty("secondary.key");
        PropertySource secondarySource = propertySourceOf("secondaryName", Pair.of(secondaryProperty.getKey(), "secondaryValue"));

        PropertyConfiguration config = configOf(primarySource, secondarySource);
        Assertions.assertEquals(Bds.mapOf(
            Pair.of(primaryProperty.getKey(), "primaryValue"),
            Pair.of(secondaryProperty.getKey(), "secondaryValue")
        ), config.getRaw());
    }
}

File: configuration/src/test/java/com/synopsys/integration/configuration/config/MultiplePropertySourceTests.kt
/**
 * configuration
 *
 * Copyright (c) 2020 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.configuration.config

import com.synopsys.integration.configuration.property.types.string.NullableStringProperty
import com.synopsys.integration.configuration.util.configOf
import com.synopsys.integration.configuration.util.propertySourceOf
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class MultiplePropertySourceTests {
    @Test
    fun primaryValueUsedOverSecondary() {
        val sharedProperty = NullableStringProperty("shared.key")
        val secondarySource = propertySourceOf("secondaryName", sharedProperty.key to "secondaryValue")
        val primarySource = propertySourceOf("primaryName", sharedProperty.key to "primaryValue")

        val config = configOf(primarySource, secondarySource);
        Assertions.assertEquals("primaryValue", config.getValue(sharedProperty))
        Assertions.assertEquals("primaryName", config.getPropertySource(sharedProperty))
    }

    @Test
    fun valueFromSecondaryWhenNotInPrimary() {
        val property = NullableStringProperty("any.key")
        val secondarySource = propertySourceOf("secondaryName", property.key to "secondaryValue")
        val primarySource = propertySourceOf("primaryName")

        val config = configOf(primarySource, secondarySource)
        Assertions.assertEquals("secondaryValue", config.getValue(property))
        Assertions.assertEquals("secondaryName", config.getPropertySource(property))
    }

    @Test
    fun containsKeysFromBothSources() {
        val primaryProperty = NullableStringProperty("primary.key")
        val primarySource = propertySourceOf("primaryName", primaryProperty.key to "primaryValue")

        val secondaryProperty = NullableStringProperty("secondary.key")
        val secondarySource = propertySourceOf("secondaryName", secondaryProperty.key to "secondaryValue")

        val config = configOf(primarySource, secondarySource)
        Assertions.assertEquals(setOf(primaryProperty.key, secondaryProperty.key), config.getKeys())
    }

    @Test
    fun rawValueMapContainsValuesFromBothSources() {
        val primaryProperty = NullableStringProperty("primary.key")
        val primarySource = propertySourceOf("primaryName", primaryProperty.key to "primaryValue")

        val secondaryProperty = NullableStringProperty("secondary.key")
        val secondarySource = propertySourceOf("secondaryName", secondaryProperty.key to "secondaryValue")

        val config = configOf(primarySource, secondarySource);
        Assertions.assertEquals(mapOf(
                primaryProperty.key to "primaryValue",
                secondaryProperty.key to "secondaryValue"
        ), config.getRaw())
    }
}